src/array/cuda/labor_sampling.cu

```diff
@@ -0,0 +1,858 @@
+          }
+        }
+        var_1 = BlockReduce(temp_storage).Sum(var_1, num_valid);
+        if (threadIdx.x == 0)
+          var_1_bcast[threadIdx.y] = var_1;
+        __syncthreads();
+        var_1 = var_1_bcast[threadIdx.y];
+
+        c *= var_1 / var_target;
+      } while (min(var_1, var_target) / max(var_1, var_target) < 1 - eps);
+
+      if (threadIdx.x == 0)
+        cs[out_row] = c;
+    }
+
+    out_row += BLOCK_CTAS;
+  }
+}
+
+template<typename IdType, int BLOCK_SIZE, IdType TILE_SIZE>
+__global__ void map_vertex_ids_global(
+  const IdType * const global,
+  IdType * const new_global,
+  const IdType num_vertices,
+  const runtime::cuda::DeviceOrderedHashTable<IdType> table) {
+  assert(BLOCK_SIZE == blockDim.x);
+
+  using Mapping = typename OrderedHashTable<IdType>::Mapping;
+
+  const IdType tile_start = TILE_SIZE * blockIdx.x;
+  const IdType tile_end = min(TILE_SIZE * (blockIdx.x + 1), num_vertices);
+
+  for (IdType idx = threadIdx.x + tile_start; idx < tile_end; idx += BLOCK_SIZE) {
+    const Mapping& mapping = *table.Search(global[idx]);
+    new_global[idx] = mapping.local;
+  }
+}
+
+}  // namespace
+
+/////////////////////////////// CSR ///////////////////////////////
+
+template <DGLDeviceType XPU, typename IdType, typename FloatType>
+std::pair<COOMatrix, FloatArray> CSRLaborSampling(CSRMatrix mat,
+                  IdArray NIDs,
+                  IdArray rows_arr,
+                  const int64_t num_picks__,
+                  FloatArray prob_arr,
+                  IdArray random_seed,
+                  IdArray cnt,
+                  int importance_sampling) {
+  constexpr bool use_csr_spmv = false;
+  constexpr bool use_hashtable = false;
+  const IdType num_picks = num_picks__;
+  const bool weights = !IsNullArray(prob_arr);
+
+  const uint64_t max_log_num_vertices = [&]() -> int {
+    for (int i = 0; i < static_cast<int>(sizeof(IdType)) * 8; i++)
+      if (mat.num_cols <= ((IdType)1) << i)
+        return i;
+    return sizeof(IdType) * 8;
+  }();
+
+  const auto& ctx = rows_arr->ctx;
+
+  workspace_memory_alloc<decltype(&ctx)> allocator(&ctx);
+
+  const auto stream = runtime::getCurrentCUDAStream();
+  const auto exec_policy = thrust::cuda::par_nosync(allocator).on(stream);
+
+  auto device = runtime::DeviceAPI::Get(ctx);
+
+  const IdType num_rows = rows_arr->shape[0];
+  IdType * rows = static_cast<IdType*>(rows_arr->data);
+  IdType * nids = NIDs->shape[0] == mat.num_cols ? static_cast<IdType*>(NIDs->data) : nullptr;
+  FloatType * A = static_cast<FloatType*>(prob_arr->data);
+
+  const IdType * const indptr = static_cast<const IdType*>(mat.indptr->data);
+  const IdType * const indices = static_cast<const IdType*>(mat.indices->data);
+  const IdType * const data = CSRHasData(mat) ?
+    static_cast<IdType*>(mat.data->data) : nullptr;
+
+  // compute in-degrees
+  auto in_deg = allocator.alloc_unique<IdType>(num_rows + 1);
+  auto cs_arr = NewFloatArray(num_rows, ctx, sizeof(FloatType) * 8);
```

Comment:
We can define a `floattype` variable and use `NDArray::Empty`.
If you prefer `NewFloatArray`, you can rewrite the `NDArray::Empty` implementation to use `NewFloatArray` instead.  Better make the code consistent.

Reply:
For IdArrays, there is a helper function NewIdArray, I wrote this to match the IdArray case.

Reply:
Got it.